### DISCUSSION TITLE
Nonsence nuber of oeprations for Fm-3m space gropup

### DISCUSSION THREAD

**Discussion by husakm** [spglib/spglib#536] [General]
I had tested a structure (C code  for generation bellow) with symprec 0.1. The spg_get_dataset gives Fm-3m group with n_operations = 1536.
This is a nonsense. Can anybody check this ? Just include my C structure description , set symprec to 0.1 and run.
```
//section  generated by checkCIF-DFT for ZZWEQ14
int num_atom = 32;
double lattice[3][3] = {
{11.5369997024536,-5.042982706982E-07,-5.12428584897862E-07},
{0,11.5369997024536,-5.1242864174128E-07},
{0,0,11.7229995727539}
};

double position[32][3] = {
{0,0,0.25},
{0,0.5,0.25},
{0,0.5,0.75},
{0.5,0.5,0.25},
{0.25,0.5,0},
{0.25,0.5,0.5},
{0.25,0,0},
{0.25,0,0.5},
{0.25,0.25,0.25},
{0.75,0.5,0},
{0.75,0.5,0.5},
{0.75,0,0},
{0.75,0,0.5},
{0.75,0.75,0.25},
{0,0,0.75},
{0.5,0,0.75},
{0.5,0,0.25},
{0.5,0.5,0.75},
{0.5,0.75,0},
{0.5,0.75,0.5},
{0,0.75,0},
{0,0.75,0.5},
{0.25,0.75,0.75},
{0.5,0.25,0},
{0.5,0.25,0.5},
{0,0.25,0},
{0,0.25,0.5},
{0.75,0.25,0.75},
{0.25,0.75,0.25},
{0.75,0.25,0.25},
{0.25,0.25,0.75},
{0.75,0.75,0.75},
};

int types[32] = {
6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,
};
//end of section  generated by checkCIF-DFT

```

**Comment by atztogo** [comment]
I have two questions for you.
1. Have you read the documentation carefully?
2. Why do you think it is nonsense?

**Reply by husakm** [reply]
1) I am not sure why is my notice related to documentation ... 

2) As far as I know  F m-3m space group number 255 have only 192 operations generating new positions, so the 1536 sounds strange.

**Reply by LecrisUT** [reply]
Just a quick analysis, `1536 / 192 = 8`. Is it possible that these are composite operations that are not included in the typical m-3m space group, e.g. a translation to `0, 0, 0.25` + other FCC operations like translation to `0.5, 0.5, 0`? Could you analyze the operations a bit?

**Reply by husakm** [reply]
The 192 operations already include the F translation, so it is de-facto 4x48. The 4x are the 0,0,0,  0,0.5.,0.5 e.t.c. additional operations generated by the F centration ....
I can not imagine 1536 operations total for any space group ...
Maybe what SPGLIB lists are not the  the space group operations but some sort of any symmetry found operations ?

**Comment by LecrisUT** [comment]
I am not an expert in crystallography so I can't judge how correct/incorrect the `1536` is or what would be the appropriate value for your structure. Could you clarify what is the expected allowed symmetries? The only thing that I might consider right now is that all of your atoms are identical so it could be that there are additional symmetry operations that could compound the basic symmetries you would expect from just the FCC structure.

**Comment by atztogo** [comment]
Space group is an infinite group, so probably you don't understand about space group.

The space group is given by $\mathrm{S}=\sum_i (W_i, w_i) \mathrm{T}$, where { $(W_i, w_i)$ } is the set of coset representatives and $\mathrm{T}$ is the translation group.

In your case, |{ $(W_i, w_i)$ }|=48. I guess you have 32 translations in your input cell. This is what spglib returns. Please read this (https://spglib.readthedocs.io/en/latest/dataset.html#rotations-translations-and-n-operations), too.

**Reply by husakm** [reply]
Space groups are groups with number of operations limited to fill 3D  periodic space. This is why they are named space groups .
For test compounds creating crystals in F m -3 m (eg. CSD code AQUINIL)  SPGLIB gives correctly 192 symmetry operations.
So there is absolute no reason why for the structure given in my sample code for the same space group the SPGLIB gives 1536 operations.
So SPGLIB identify space group incorrectly, or the number of operation does not correspond to the given group.

**Reply by atztogo** [reply]
I expect the output to be 1536, which aligns with how spglib is implemented. If you find this result unreasonable, then perhaps spglib may not be the best tool for your purpose. You can close this discussion.

**Reply by husakm** [reply]
It is not clear to me what spglib returns as symmetry operations:
A) The symmetry operations of the space group witch it had identifed for given data
B) Something else.
I am not sure what B) should be an how it is defined. It is probably not what is listed in international tables for X-ray crystallography as individual space group properites ....

**Reply by atztogo** [reply]
We wrote a paper (https://www.tandfonline.com/doi/full/10.1080/27660400.2024.2384822) to answer your question. Please read it.

**Reply by husakm** [reply]
Dear Mr. Atsushi Togo

I did read the actual version of the article ( I had read the old version in the past as well) ...
Notices to the article :

1) In the list of the the software doing similar things as spglib you shoudl probabaly mention the most often used code by all crystallographers available free as a Fortran code . PLATON and its ADDSYM module:
https://www.platonsoft.nl/platon/pl000401.html . It is a part of the CheckCif system for crystal structures vlaidation.

2) In the list of software utilizing spglib it will be worth to add one of the most important solid state QM programs operating with plane waves: CASTEP . Btw. the use of spglib in Avogradro is done in really wrong way - Avogrador can not work generally correctly with space groups and symmetry . Adding of SPGLIG to Avogadro gives limited sense ...

According the topic discussed from begging - why can you not simply tell the spglib returns all operations witch it had found and not the operations corresponding to the identified  space group ?
Can you tell me what will be the maximum number of such operations (it is impossible to find this anywhere) . I need to know this number for more safe interfacing between our C# code and spglib (memory allocation). I hope it is not infinity as you mention in one of your answers.

Sincerely Yours
Michal Husak
Department of Solid State Chemistry
UCT Prague
assoc. professor.

**Reply by atztogo** [reply]
> why can you not simply tell the spglib returns all operations witch it had found and not the operations corresponding to the identified space group ?

I don't see what you mean.

> Can you tell me what will be the maximum number of such operations (it is impossible to find this anywhere)

Possible maximum number is calculated by multiplying the number of atoms in the input structure by 48.

I am feeling quite tired. If I become too exhausted, I won’t be able to continue developing spglib for users. Please understand that I am not obligated to answer your question.

**Comment by LecrisUT** [comment]
Regarding the discussion, I believe the issue is that of interfacing. In one case you are interogating the Fm-3m structure and asking it what symmetry operations it has, in which case @husakm would expect `192`. In the other case you are interogating  the lattice structure itself and you get all the possible symmetry operations in which case it can be `n * 192` depending on the atom types which is what @atztogo implemented afaiu, and what you would want from the system. Whether or not the system in question is truly Fm-3m, or if there is nuance to the atoms in a Fm-3m structure, I do not have the vocabulary or expertise to judge, but having simple example of one of the operations that is not in a typical Fm-3m structure, that would help clear out the situation more.

> 1\. In the list of the the software doing similar things as spglib you shoudl probabaly mention the most often used code by all crystallographers available free as a Fortran code
> 2\. In the list of software utilizing spglib it will be worth to add one of the most important solid state QM programs operating with plane waves: CASTEP

By all means, do make a PR to include what software you find to be popular uses. 

> Can you tell me what will be the maximum number of such operations (it is impossible to find this anywhere) .

That is a reasonable request, and I believe we hit this issue in the tests as well at some point @atztogo. (Edit: ah, you've just answered in the previous part)

> I need to know this number for more safe interfacing between our C# code and spglib (memory allocation).

I want to revisit packaging for nuget. And I will start having time for projects now that my previous contract finished. It would be nice if we can discuss this point again. Also note that spglib has been packaged for `vcpkg` which I hope would make it easier for Windows users, let me know if there are other environments where it might be useful to distribute.

**Reply by atztogo** [reply]
@LecrisUT, we also have https://spglib.readthedocs.io/en/latest/api.html#spg-get-multiplicity.

**Reply by LecrisUT** [reply]
> @LecrisUT, we also have https://spglib.readthedocs.io/en/latest/api.html#spg-get-multiplicity.

Wouldn't that basically double the operations performed? Once to get the number of operations and once to save them in the output? On the other hand, if you get `SpglibDataset` directly, doesn't spglib already populate both the number of operations and the array of operations?

**Reply by atztogo** [reply]
You are absolutely correct. However, this function may still be useful for those who don’t mind the additional operations. For me, `spg_get_dataset` is fine but it seems not for @husakm.

**Reply by LecrisUT** [reply]
I think it's more of an issue of communication. We have just recently started to mark interfaces as deprecated and I believe there are still other interfaces that are vague if they need to be supported and exposed or not, and @husakm is mostly implementing an over-complete interface. I wish to discuss with him more on that front and to iron out a cohesive interface across all OOP languages. I still want to revisit the restructuring in #489 which would help simplify a lot of these questions and misunderstandings.

**Reply by husakm** [reply]
Generally I develop a complete C#/spglib interface able to get any data SPGLIB generate. From technical/safety reason witch I was not able to overcame, I declare memory areas inside C# for data storage witch SPGLIB generate (C# Marshaling). I do no know in advance the number of rotation/translation spglib generate so I had declared 192 elements array for translations and rotations as the maximum giving sense for this interface. The code crashed when the 1536 operations had overwritten other parts of memory ....
I can modify the C code of spglib myself, add an interface not operating with the array of translations and rotations with have (as I see now) not pre-defined maximal size. I was just curios what is the source of this non usual operations number.
